File: DigitalImageProcessing/cpp/gauss_kernel.cpp
/**
 * @File    : gauss_kernel.cpp
 * @Brief   : 计算高斯卷积算子；高斯卷积核的可分离；高斯卷积核进行图像平滑(模糊)
 * @Author  : Wei Li
 * @Date    : 2021-09-17
*/

#include <iostream>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>

cv::Mat sepConv2D_Y_X(const cv::Mat &image, cv::Mat gaussKernel_y, cv::Mat gaussKernel_x)
{
    // 可分离高斯卷积
    cv::Mat sep_gauss_img_y, sep_gauss_img;
    // 然后利用 OpenCV 函数进行离散卷卷积操作(该函数只是相关计算，而不是卷积操作)
    cv::filter2D(image, sep_gauss_img_y, -1, gaussKernel_y, cv::Point(0, 0), 0.0, cv::BORDER_CONSTANT);
    cv::filter2D(sep_gauss_img_y, sep_gauss_img, -1, gaussKernel_x, cv::Point(0, 0), 0.0, cv::BORDER_CONSTANT);

    return sep_gauss_img;
}

cv::Mat gaussBlur(const cv::Mat &image, cv::Size winSize, float sigma,
                  int ddepth = CV_64F, cv::Point anchor = cv::Point(-1, -1), int borderType = cv::BORDER_DEFAULT)
{
    // 高斯卷积核为奇数
    CV_Assert(winSize.width % 2 == 1 && winSize.height % 2 == 1);
    // 垂直方向高斯卷积核
    cv::Mat gaussKernel_y = cv::getGaussianKernel(sigma, winSize.height, CV_64F);
    // 水平方向 and transpose
    cv::Mat gaussKernel_x = cv::getGaussianKernel(sigma, winSize.width, CV_64F);
    gaussKernel_x = gaussKernel_x.t(); // 转置
    // 可分离的高斯卷积核,自定义函数
    cv::Mat blurImgae = sepConv2D_Y_X(image, gaussKernel_y, gaussKernel_x);

    return blurImgae;
}

// -------------------------------
int main(int argc, char **argv)
{
    if (argc > 1)
    {
        cv::Mat image = cv::imread(argv[1], 0);
        if (!image.data)
        {
            std::cout << "Error: reading image unsuccesfully." << std::endl;
            return -1;
        }
        cv::imshow("OriginImage", image);

        // 1. 自定义高斯卷积
        cv::Mat first_gauss_blur_img = gaussBlur(image, cv::Size(7, 7), 1.5);
        cv::imshow("firstGaussBlurImg", first_gauss_blur_img);

        // 2. OpenCV 函数进行高斯卷积
        cv::Mat second_gauss_blur_img;
        cv::GaussianBlur(image, second_gauss_blur_img, cv::Size(7, 7), 1.5, 1.5);
        cv::imshow("secondGaussBlurImg", second_gauss_blur_img);

        cv::waitKey(0);
        cv::destroyAllWindows();
        return 0;
    }
    else
    {
        std::cout << "Usage: OpenCV gauss convolution image." << std::endl;
        return -1;
    }
}

File: DigitalImageProcessing/python/gauss_kernel.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Funciton: 计算高斯卷积算子；高斯卷积核的可分离；高斯卷积核进行图像平滑(模糊)
@Python Version: 3.8
@Author: Wei Li
@Date: 2021-09-17
"""

import math
import sys
from scipy import signal
import numpy as np
import cv2 as cv


def getGaussKernel(sigma, H, W):
    """计算高斯卷积算子。

    Args:
        sigma (float): 高斯分布的标准差 sigma
        H (int): 高斯核的大小，奇数
        W (int): 高斯核的大小，奇数

    Returns:
        [ndarray]: 高斯卷积核算子
    """
    # 1. 构建高斯矩阵
    # -------------------------
    gaussMatrix = np.zeros([H, W], np.float32)
    # 计算中心点
    centerH = (H - 1) / 2
    centerW = (W - 1) / 2
    for idx_row in range(H):
        for idx_col in range(W):
            norm2 = math.pow(idx_row - centerH, 2) + math.pow(idx_col - centerW, 2)
            gaussMatrix[idx_row][idx_col] = math.exp(-norm2 / (2 *math.pow(sigma, 2)))
    # 利用 Numpy 进行简化
    # gaussMatrix = np.exp(-0.5 * (np.power(idx_row) + np.power(idx_col)) / math.pow(sigma, 2))
    # -------------------------

    # 2. 计算高斯矩阵的和
    sumGM = np.sum(gaussMatrix)
    # 3. 归一化高斯矩阵， 即得到高斯卷积算子
    gaussKernel = gaussMatrix / sumGM

    return gaussKernel


def gaussBlur(image, sigma, H, W, mode, _boudary="fill", _fillvalue=0):
    # 水平方向高斯卷积核
    gaussKernel_x = cv.getGaussianKernel(sigma, W, cv.CV_64F)
    # 转置
    gaussKernel_x = np.transpose(gaussKernel_x)
    # 图像矩阵与水平高斯核卷积
    gaussBlur_x = signal.convolve2d(image, gaussKernel_x, mode=mode, boundary=_boudary, fillvalue=_fillvalue)
    # 构建垂直方向的高斯卷积核
    gaussKernel_y = cv.getGaussianKernel(sigma, H, cv.CV_64F)
    # 与垂直方向上的高斯核卷积
    gaussKernel_xy = signal.convolve2d(gaussBlur_x, gaussKernel_y, mode=mode, boundary=_boudary, fillvalue=_fillvalue)

    return gaussKernel_xy

    
# --------------------------
if __name__ == "__main__":
    # 1. 计算高斯卷积算子
    sigma = 1.2
    H, W = 3, 3
    gaussKernel = getGaussKernel(sigma=sigma, H=H, W=W)
    print(f"The gauss kernel of 3x3 is:\n{gaussKernel}")

    # 2. 高斯卷积算子是可分离的，只需要一个方向(垂直或者水平)的 API即可
    gauss_kernel_v = cv.getGaussianKernel(3, 1.2, cv.CV_64F)  # 垂直方向
    print(f"The vertical direction of Gaussian Kernel:\n{gauss_kernel_v}")
    # 对垂直方向的 kernel 进行转置
    print(f"The horizontal direction of Gaussian Kernel:\n{gauss_kernel_v.T}")
    # 验证一下，高斯卷积算子是可分离的
    
    # 3. 利用 高斯卷积核进行图像平滑
    if len(sys.argv) > 1:
        image = cv.imread(sys.argv[1], 0)
        if image is None:
            print(f"Error: no such file or directory.")
            sys.exit()
        cv.imshow("OriginImage", image)
        blurImage = gaussBlur(image, 5, 51, 51, "same")
        # 对 blurImage 进行灰度级显示，float to uint8
        blurImage = (np.round(blurImage)).astype(np.uint8)
        cv.imshow("GaussBlurImage", blurImage)

        cv.waitKey(0)
        cv.destroyAllWindows()

    else:
        print(f"Usage: python gaussBlur imageFile.")
